### ISSUE TITLE
Provider credentials not refreshed on several command paths

### ISSUE THREAD

**Issue by fullykubed** [pulumi/pulumi#4981]
`pulumi` CLI:  v2.5.0
`@pulumi/vault`: v2.2.1
`@pulumi/aws`: v2.11.0

We have a setup where we retrieve credentials from external resources as configuration to our pulumi providers. Example below:

```Typescript
import * as aws from "@pulumi/aws";
import * as vault from "@pulumi/vault";
import { config } from "./config";

export const awsCfg = vault.aws
  .getAccessCredentials({
    backend: "aws",
    role: config.require('awsRole'),
    type: "sts",
  })
  .then((creds) => ({ ...creds, region: config.require('awsRegion') }));

export const awsProvider = new aws.Provider("default", {
  accessKey: awsCfg.then(({ accessKey }) => accessKey),
  secretKey: awsCfg.then(({ secretKey }) => secretKey),
  token: awsCfg.then(({ securityToken }) => securityToken),
  region: awsCfg.then(({ region }) => region as aws.Region),
});
```

However, those credentials get stored in the state file for the stack and appear to be used for resource deletion even though they are dynamic. This results in invalid API requests when destroying resources. See [logs](https://pastebin.com/Wf5cbx2i).

A couple of things to propose:

- Currently, when the AWS provider has stale credentials, the CLI appears to hang on destroy; it was only by enabling tracing that we were able to figure out what was wrong. It would be nice if the CLI was a little more descriptive in the case of invalid creds.

- The [troubleshooting docs](https://www.pulumi.com/docs/troubleshooting/) or the [migrating from terraform](https://www.pulumi.com/docs/guides/adopting/from_terraform/) docs might want to have an expanded section on when to use refresh (e.g., updating dynamic creds, ensuring accurate diff, etc.) as I imagine this could definitely confuse other users coming from terraform (the setup depicted above is a direct translation of one of our terraform files and was basically our first experience with Pulumi).

- This is another reason why automatic refresh as proposed in issue #2247 may be a good idea; at least, it would be nice to be able to mark specific resources as needing to be refreshed on each run

Thanks for hearing me out!

**Comment by jclangst** [comment]
The problem with the above configuration seems to be even deeper than I originally thought. It appears that `up --refresh` exhibits the same behavior as `destroy`... the cli hangs and inspecting the trace logs shows that it is attempting to use the old credentials for the resources, and does not actually refresh the provider credentials. I think this is especially confusing as one would expect `--refresh` to get the new credentials.

Currently, the only workaround that I have found is to run `up` to get the latest credentials into the statefile and then run whatever subsequent commands depends on them. I feel that there should be another more elegant approach for this situation; is there something that we are missing?

**Comment by jclangst** [comment]
For reference, this is a simplified CI script that I have whipped up in order to make sure that the `--refresh` flag does not cause failures:

```bash
#!/usr/bin/env bash

set -e
ENV=${1:-devops}

# update the credentials
pulumi up --target "urn:pulumi:auth-service.$ENV::auth-service::pulumi:providers:aws::default" -s "auth-service.$ENV" --skip-preview --non-interactive --yes

# refresh and update all of the other resources
pulumi up -s "auth-service.$ENV" --refresh --skip-preview --non-interactive --yes
```

**Comment by jclangst** [comment]
Just adding that the same issue occurs when running `up --target <urn>` on a resource that uses that uses a provider with the stale creds.

**Comment by sy-bee** [comment]
this is a breaking issue when credentials gets rotated and you try to run refresh on resources created with previous credentials.

**Comment by KDanisme** [comment]
I am getting the same issue, with different providers and the latest pulumi version.

**Comment by twoTimesAgnew** [comment]
Had this issue as well on credential rotation. The workaround for now seems to be to delete all references of the `provider` from the stack file (`pulumi stack export` then `pulumi stack import`), then doing a `pulumi refresh`.

**Comment by michal-bajer1** [comment]
Has there been any progress?
I have run into a different version of the same issue. 
Azure-native provider has a flag called `useOidc`, which is true or false. You can only set it to `true` in environments, which support OIDC, but it gets baked into the state.
Which means, if you for example:
1. Run `pulumi up` locally and therefore set it to `false` based on the lack of ARM_USE_OIDC variable (which is what should actually govern this behavior).
2. You will never be able to run refresh in a pipeline (even though ARM_USE_OIDC is set to true), because the false value is baked into the state and refresh ignores the environment variable.
3. Refresh also ignores values you are passing into the object, which means there is no way to change the value for refresh execution until you set the value in the state to `true` value somewhere.

Is there any logical reason why explicitly created providers behave completely different to the default provider when authenticating?

**Comment by algompluecker** [comment]
I also face a related issue when using update plans. I am using OIDC tokens which are short lived and create a plan for review. Once the plan is approved, it shall be applied using `pulumi up --plan=plan.json`.
Since the token has expired during the review, we need to get a fresh token for the up command. But since the provider secrets are stored in the state / plan, this results in an error: `...pulumi:providers:azure-native::myProvider violates plan: properties changed: ++oidcToken[{***}!={***}]`

I also think provider secrets should not be persisted in the state / plan or there should be a CustomResourceOption to exclude them.

**Comment by pgavlin** [comment]
Cross-posting from pulumi/esc#199: https://github.com/pulumi/esc/issues/199#issuecomment-2164123853

**Comment by mtparet** [comment]
This issue prevents to use Pulumi with multiple instances of a same provider within the same stack using temporary secrets. This is a blocker to allow more adoption within our current project.

**Comment by lukehoban** [comment]
One likely path for addressing this issue is https://github.com/pulumi/pulumi/issues/16600, which would transition to running the user program on `pulumi refresh` and `pulumi destroy` operations.  This would ensure that new values get passed into provider inputs at the time of the new operation, instead of just always using whats in state.  There may be other ways we could provide mitigations for this issue, but fundamentally, the natural user expectation is that their code is run to compute the new inputs to explicit providers, even on `refresh` and `destroy` operations.

**Comment by lunaris** [comment]
I _think_ that the use cases listed here are now addressed by the recent support for `--run-program`:

* `pulumi refresh --run-program` will run the program when refreshing, allowing for credentials to be picked up
* `pulumi destroy --run-program` will do the same when destroying
* `pulumi [preview|up] --refresh --run-program` should run the program prior to the refresh operations triggered by `refresh`.

I'm thus going to optimistically mark this as fixed, but please reopen/raise another issue if there's a path that doesn't work for you still 🙏
